### ISSUE TITLE
Improper merging of unequal keys in map-merge

### ISSUE THREAD

**Issue by davidkpiano** [sass/libsass#1283]
**TEST:**
```scss
$map: ();

$map: map-merge($map, (1 2: 3));
$map: map-merge($map, (2 1: 3));

.test {
  test: inspect($map);
}
```

**EXPECTED:** (Ruby Sass)
```scss
.test {
  test: (1 2: 3, 2 1: 3);
}
```

**ACTUAL:** (LibSass 3.2.5)
```scss
.test {
  test: (1 2: 3);
}
```

I'll write specs for all these issues soon, I promise!

**Comment by mgreter** [comment]
Thxs for the report @davidkpiano! I can confirm this issue and also have a fix ready!

In ast.hpp `class List` `virtual size_t hash()`:
```c++
- hash_ ^= (elements()[i])->hash();
+ hash_ ^= (elements()[i])->hash() + i;
```

**Comment by mgreter** [comment]
First solution approach was too naive, should use this instead:

```
  template <typename T>
  void hash_combine (std::size_t& seed, const T& val)
  {
    seed ^= std::hash<T>()(val) + 0x9e3779b9
             + (seed<<6) + (seed>>2);
  }
...
      for (size_t i = 0, L = length(); i < L; ++i)
        hash_combine(hash_, (elements()[i])->hash());
```

**Comment by xzyfer** [comment]
TODO: link the above hash_combine function to it's source at http://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html

**Comment by xzyfer** [comment]
Specs added https://github.com/sass/sass-spec/pull/414

**Comment by okdana** [comment]
Thank you for fixing this!

**Comment by xzyfer** [comment]
This is fixed and will be in 3.3.
